[SymmMem] Experimental NVSHMEM integration

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151993
* #151819
* #151498
* __->__ #151261

Adding NVSHMEM as a backend for `SymmetricMemory`, implementation of which is in `NVSHMEMSymmetricMemory.cu`.

Moving some helper functions in `CUDASymmetricMemory.cu` to `CUDASymmetricMemoryUtils.cpp`, so that they can be shared by `NVSHMEMSymmetricMemory`. These functions are mostly side-band exchange helpers (`store_all_gather`, `IpcChannel`, etc).

Adding `TORCH_SYMMEM` to control which implementation to use for CUDA tensors, currently support: `CUDA` (in-house impl), `NVSHMEM`.

The NVSHMEM feature is gated by build-time flag: `USE_NVSHMEM=1`. And `NVSHMEM_HOME` setting is required (TODO).

Ported most code from #146593.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 committed:
[WIP] experimental NVSHMEM integration

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/nvshmem_extension.cu:
```diff
@@ -0,0 +1,421 @@
+    T val) {
+  static_assert(std::is_trivially_copyable_v<T>);
+
+  std::vector<std::string> peer_keys;
```
Call reserve

Skylion007 commented on torch/csrc/distributed/c10d/nvshmem_extension.cu:
```diff
@@ -0,0 +1,421 @@
+
+  std::vector<std::string> peer_keys;
+  for (int r = 0; r < world_size; ++r) {
+    std::ostringstream oss;
```
May be faster / cleaner with fmtlib

Skylion007 commented on torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:
```diff
@@ -916,15 +917,354 @@ c10::intrusive_ptr<Block> CUDASymmetricMemoryAllocator::find_block(void* ptr) {
+  NVSHMEMSymmetricMemory(
+      std::shared_ptr<NVSHMEMAllocation> allocation,
+      const std::string& group_name)
+      : allocation_(allocation),
```
Move the shared pointer into allocation_, have the other members call the methods on "allocation_"

kwen2501 committed:
Update on "[WIP] experimental NVSHMEM integration"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP] experimental NVSHMEM integration"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP] experimental NVSHMEM integration"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[SymmMem] Experimental NVSHMEM integration"


Adding NVSHMEM as a backend for `SymmetricMemory`, implementation of which is in `NVSHMEMSymmetricMemory.cu`.

Moving some helper functions in `CUDASymmetricMemory.cu` to `CUDASymmetricMemoryUtils.cpp`, so that they can be shared by `NVSHMEMSymmetricMemory`. These functions are mostly side-band exchange helpers (`store_all_gather`, `IpcChannel`, etc).

Adding `TORCH_SYMMEM` to control which implementation to use for CUDA tensors, currently support: `CUDA`, `NVSHMEM`.

The NVSHMEM feature is gated by build-time flag: `USE_NVSHMEM=1`. And `NVSHMEM_HOME` setting is required (TODO).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented on caffe2/CMakeLists.txt:
```diff
@@ -976,6 +977,41 @@ elseif(USE_CUDA)
     target_link_libraries(torch_cuda PRIVATE __caffe2_nccl)
     target_compile_definitions(torch_cuda PRIVATE USE_NCCL)
   endif()
+
+  # Use env var for these for now for prototyping purposes
+  set(USE_NVSHMEM $ENV{USE_NVSHMEM} CACHE BOOL "Enable NVSHMEM support")
+  set(NVSHMEM_HOME $ENV{NVSHMEM_HOME} CACHE PATH "Path to NVSHMEM build dir")
+
+  if(USE_NVSHMEM)
+    set(NVSHMEM_INCLUDE_DIR "${NVSHMEM_HOME}/include")
+    set(NVSHMEM_LIB_DIR "${NVSHMEM_HOME}/lib")
+
+    include_directories(${NVSHMEM_INCLUDE_DIR})
+
+    # Linking with nvshmem requires the source binary to be built with -rdc
+    # which is not viable for libtorch_cuda. So we isolate the linking of
+    # nvshmem in nvshmem_extension.
+    add_library(nvshmem_extension SHARED
+        "${TORCH_SRC_DIR}/csrc/distributed/c10d/nvshmem_extension.cu"
+        "${TORCH_SRC_DIR}/csrc/distributed/c10d/NVSHMEMSymmetricMemory.cu"
+        "${TORCH_SRC_DIR}/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp"
+        "${TORCH_SRC_DIR}/csrc/distributed/c10d/cuda/utils.cpp"
+    )
+    set_target_properties(nvshmem_extension PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
+    target_compile_options(nvshmem_extension PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-rdc=true>)
+    target_compile_options(nvshmem_extension PRIVATE "-U__CUDA_NO_HALF_OPERATORS__")
+    target_link_libraries(nvshmem_extension PRIVATE
+        ${NVSHMEM_LIB_DIR}/libnvshmem.a
+        ${NVSHMEM_LIB_DIR}/nvshmem_bootstrap_uid.so
+    )
+    target_link_libraries(nvshmem_extension PRIVATE mlx5)
+    target_link_libraries(torch_cuda PRIVATE nvshmem_extension)
+    install(TARGETS nvshmem_extension EXPORT Caffe2Targets DESTINATION lib)
+    install(
+        FILES ${NVSHMEM_LIB_DIR}/nvshmem_bootstrap_uid.so
+        DESTINATION lib
+    )
+  endif()
```
Hi @atalman @malfet do you mind having a look at this section? Today we gate the NVSHMEM feature with build-time flag `USE_NVSHMEM=1`. We may change it later.

kwen2501 committed:
Update on "[SymmMem] Experimental NVSHMEM integration"


Adding NVSHMEM as a backend for `SymmetricMemory`, implementation of which is in `NVSHMEMSymmetricMemory.cu`.

Moving some helper functions in `CUDASymmetricMemory.cu` to `CUDASymmetricMemoryUtils.cpp`, so that they can be shared by `NVSHMEMSymmetricMemory`. These functions are mostly side-band exchange helpers (`store_all_gather`, `IpcChannel`, etc).

Adding `TORCH_SYMMEM` to control which implementation to use for CUDA tensors, currently support: `CUDA`, `NVSHMEM`.

The NVSHMEM feature is gated by build-time flag: `USE_NVSHMEM=1`. And `NVSHMEM_HOME` setting is required (TODO).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fegin commented on torch/csrc/distributed/c10d/SymmetricMemory.cpp:
```diff
@@ -269,6 +274,10 @@ TORCH_LIBRARY_FRAGMENT(symm_mem, m) {
+
+  m.def("nvshmem_broadcast(Tensor(a!) input, str group_name) -> Tensor(a!)");
+  m.def(
+      "nvshmem_reduce_scatter_out(Tensor input, str group_name, Tensor(a!) out) -> Tensor(a!)");
```
Why do these two APIs have the `nvshmem_` prefix? Do we plan to unify `nvshmem_reduce_scatter_out` with `reduce_scatter_out`?

kwen2501 replied:
Correct. I expect we can remove the prefix by the release.

fegin commented on torch/csrc/distributed/c10d/nvshmem_extension.cu:
```diff
@@ -0,0 +1,390 @@
+  dim3 grid_dim(32), block_dim(544);
+  auto stream = at::cuda::getCurrentCUDAStream();
+  nvshmemx_barrier_on_stream(team, stream);
+  nvshmemx_collective_launch(
```
Just for my learning, we need to use `nvshmemx_collective_launch` because `nvshmem_all_reduce_kernel` uses some nvshmem APIs. Is this correct?

kwen2501 replied:
From documentation:
> The nvshmemx_collective_launch function must be used to launch CUDA kernels on the GPU when the CUDA kernels use NVSHMEM synchronization or collective APIs (e.g., nvshmem_wait, nvshmem_barrier, nvshmem_barrier_all, or any other collective operation). CUDA kernels that do not use synchronizing NVSHMEM APIs (or that do not use NVSHMEM APIs at all), are not required to be launched by this API. This call is collective across the PEs in the NVSHMEM job.

kwen2501 commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -37,26 +37,6 @@ def enable_symm_mem_for_group(group_name: str) -> None:
         f"symmetric_memory-{global_ranks_str}",
         c10d._get_process_group_store(group),
     )
-    # Use one store-based broadcast to bootstrap a file store from the process
```
I guess @yifuwang removed it for two reasons:
(1) yes, to support cross-node scenario;
(2) it is indeed not needed even for `CUDASymmetricMemory` because the special handle exchange is now handled by `IpcChannel`.

kwen2501 committed:
Update on "[SymmMem] Experimental NVSHMEM integration"


Adding NVSHMEM as a backend for `SymmetricMemory`, implementation of which is in `NVSHMEMSymmetricMemory.cu`.

Moving some helper functions in `CUDASymmetricMemory.cu` to `CUDASymmetricMemoryUtils.cpp`, so that they can be shared by `NVSHMEMSymmetricMemory`. These functions are mostly side-band exchange helpers (`store_all_gather`, `IpcChannel`, etc).

Adding `TORCH_SYMMEM` to control which implementation to use for CUDA tensors, currently support: `CUDA` (in-house impl), `NVSHMEM`.

The NVSHMEM feature is gated by build-time flag: `USE_NVSHMEM=1`. And `NVSHMEM_HOME` setting is required (TODO).

Ported most code from #146593.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp:
```diff
@@ -0,0 +1,223 @@
+      true;
+}
+
+// Query environment variable to get the backend used for CUDA Symmetric Memory.
```
what's your plan for documenting this to users?
a) how long-lived is it? (iiuc, it's long-lived since we have to support 'Cuda' backend for AMD
b) what are the user-facing tradeoffs?  Why would i choose CUDASymmetricMemory vs NVSHMEMSymmetricMemory in a given scenario?

kwen2501 replied:
a) The env itself could be longer term due to possibility of "NCCL" backend. The "CUDASymmetricMemory" impl could be shorter-termed but still useful for the reason you mentioned.
b) intra vs inter capability is the main difference from user point of view.

wconstab replied:
(b) is still not clear to me. Does nvshem backend support intra+inter or is it purely intra?

if nvshmem supports the superset of cuda backend _except_ for AMD support, then I would argue we delete the env, and autodetect AMD to switch it to the cuda backend. otherwise, default to the nvshmem backend?

kwen2501 replied:
The env is also for NCCLSymmetricMemory vs NVSHMEMSymmetricMemory.

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp:
```diff
@@ -0,0 +1,223 @@
+  auto socket_name = get_socket_name(dst_pid);
+  std::copy(socket_name.begin(), socket_name.end(), addr.sun_path);
+
+  struct iovec io = {.iov_base = (void*)("fd"), .iov_len = 2};
```
what is this (void*)("fd") thing?

huydhn commented:
@pytorchbot drci

wconstab commented on caffe2/CMakeLists.txt:
```diff
@@ -976,6 +977,41 @@ elseif(USE_CUDA)
+    # which is not viable for libtorch_cuda. So we isolate the linking of
+    # nvshmem in nvshmem_extension.
+    add_library(nvshmem_extension SHARED
+        "${TORCH_SRC_DIR}/csrc/distributed/c10d/nvshmem_extension.cu"
```
i was gonna suggest having a dedicated folder for the nvshmem backend files so its clear to readers that they are built into a separate compilation unit.

however maybe its an annoyingly long path? `csrc/distributed/c10d/symm_mem/common` and `/nvshmem`

kwen2501 replied:
Yeah, totally agree with a dedicated folder. The c10d folder is a mix of everything today, will refactor together with other libraries later.

fduwjj replied:
maybe we can have a symmetric memory dedicated folder?

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:
```diff
@@ -1,275 +1,32 @@
 #define CUDART_SUPPORTS_MULTICAST
 #endif
 
-namespace {
```
you should really do the code-motion part in a previous PR so its easier to tell what is real changes

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:
```diff
@@ -1,275 +1,32 @@
 
-static size_t store_comm_seq_id = 0;
-
-template <typename T>
```
3 functions moved to utils file; not checking the details, assuming its just a direct move

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp:
```diff
@@ -0,0 +1,223 @@
+namespace c10d::symmetric_memory {
+
+bool device_has_multicast_support(int device_idx) {
+  if (c10::utils::check_env("TORCH_SYMM_MEM_DISABLE_MULTICAST") == true) {
```
planning to document these ENVs? any idea if they are going to be long-lived?

kwen2501 replied:
We can probably clean this env.

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:
```diff
@@ -1,275 +1,32 @@
+/* Start of CUDASymmetricMemory implementation */
 
-constexpr size_t signal_pad_size = 2048;
 const std::string store_comm_prefix = "CUDASymmetricMemory";
```
just curious, since this is declared as 'extern' in the header, what does that mean when there is a definition in both torch and nvshmem backend?  won't the torch definition always exist and take precedence even when we are using nvshmem backend?  (did you confirm in tests that the nvshmem backend prefix is being used)

kwen2501 replied:
@wconstab thanks for your suggestion. I made a little class `StoreExchange` to encapsulate those `store_` methods and the prefix. Then each SymmMem implementation can locally instantiate a `StoreExchange` instance to avoid namespace & prefix contention.

wconstab commented on torch/csrc/distributed/c10d/SymmetricMemory.cpp:
```diff
@@ -212,15 +213,19 @@ namespace {
     const at::Tensor& local_input,
+    // NOLINTNEXTLINE(performance-unnecessary-value-param)
     std::string reduce_op,
+    // NOLINTNEXTLINE(performance-unnecessary-value-param)
```
nit: would it be clearer to do `(void) reduce_op` in the meta kernel? its not a big deal though

kwen2501 replied:
I think dispatcher would complain about the signature mismatch.

wconstab replied:
no, you keep the signature.

in theory this is the preferred way to tell the compiler/linter that you're intentionally not using a provided value and iiuc it should silence linter errors, but .. your way is fine too
```
void foo(int x){
   (void)x;
   return;
}
```

kwen2501 replied:
Ah I see what you mean now. Thanks for the suggestion!
In this specific case, the linter wants me to change `std::string` to `const std::string&` to avoid a copy in argument passing. But `const std::string&` is unfortunately not supported by torch dispatcher signature definition (if I am not mistaken).

wconstab commented on torch/csrc/distributed/c10d/SymmetricMemory.hpp:
```diff
@@ -71,6 +71,14 @@ class TORCH_API SymmetricMemory : public c10::intrusive_ptr_target {
   virtual int get_rank() = 0;
   virtual int get_world_size() = 0;
+
+  virtual std::vector<int> get_rank_to_global_rank() {
```
is this NYI or NotImplemented?  (Will we implement it for Cuda backend?  should this be documented?  (why is this different than c10d existing rank mapping helpers)

kwen2501 replied:
Yes, it stands for unimplemented. 
Good question whether other backends need it. Perhaps I should move them into `NVSHMEMSymmetricMemory` for now.

wconstab commented on torch/csrc/distributed/c10d/CUDASymmetricMemoryUtils.cpp:
```diff
@@ -0,0 +1,223 @@
+  }
+}
+
+IpcChannel::IpcChannel()
```
i didn't review the IpcChannel code yet, maybe someone else can take a close look there

kwen2501 replied:
It is a move. Sorry

fegin commented on torch/csrc/distributed/c10d/nvshmem_extension.cu:
```diff
@@ -0,0 +1,123 @@
+  return team;
+}
+
+at::Tensor nvshmem_broadcast(at::Tensor& input, const std::string& group_name) {
```
one minor nit: the CUDA backend puts the ops (e.g., one_shot_all_reduce) in CUDASymmetricMemoryOps.cu. Do we want to follow that convention?

kwen2501 commented on torch/csrc/distributed/c10d/SymmetricMemory.cpp:
```diff
@@ -269,6 +274,8 @@ TORCH_LIBRARY_FRAGMENT(symm_mem, m) {
   m.def(
       "memset32_(Tensor(a!) input, int offset, int val, int count) -> Tensor(a!)");
+
+  m.def("nvshmem_broadcast(Tensor(a!) input, str group_name) -> Tensor(a!)");
```
The latter I guess.

fegin reviewed (approved):
We can do the APIs renaming, file moving in latter PRs. Please fix the linter error before landing.

kwen2501 commented:
@fegin Thanks. No lint error AFAIK.

kwen2501 commented:
@pytorchbot merge

fduwjj reviewed (approved):
I think in general this looks good but we really want to separate refactor and new code. Landing a large PR will increase the risk of your PR getting reverted which happens many times this half.

kwen2501 committed:
Update on "[SymmMem] Experimental NVSHMEM integration"


Adding NVSHMEM as a backend for `SymmetricMemory`, implementation of which is in `NVSHMEMSymmetricMemory.cu`.

Moving some helper functions in `CUDASymmetricMemory.cu` to `CUDASymmetricMemoryUtils.cpp`, so that they can be shared by `NVSHMEMSymmetricMemory`. These functions are mostly side-band exchange helpers (`store_all_gather`, `IpcChannel`, etc).

Adding `TORCH_SYMMEM` to control which implementation to use for CUDA tensors, currently support: `CUDA` (in-house impl), `NVSHMEM`.

The NVSHMEM feature is gated by build-time flag: `USE_NVSHMEM=1`. And `NVSHMEM_HOME` setting is required (TODO).

Ported most code from #146593.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
